### PR TITLE
make secret types idempotent

### PR DESCRIPTION
**Pull request by Atheuz**
## Change Summary

make secret types idempotent

## Related issue number

#1044

## Checklist

* [X] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [X] Documentation reflects the changes where applicable
* [X] `changes/1330-atheuz.md`

### CODE REVIEW

File: pydantic/types.py [hunk 1]
```diff
@@ -530,11 +530,13 @@ def __modify_schema__(cls, field_schema: Dict[str, Any]) -> None:
 
     @classmethod
     def __get_validators__(cls) -> 'CallableGenerator':
-        yield str_validator
         yield cls.validate
 
     @classmethod
-    def validate(cls, value: str) -> 'SecretStr':
+    def validate(cls, value: Any) -> 'SecretStr':
+        if value.__class__ == cls:
```
Comment (samuelcolvin) [inline pydantic/types.py:537]: why not isinstance?

File: docs/examples/types_secret_types.py [hunk 2]
```diff
@@ -3,19 +3,27 @@
 class SimpleModel(BaseModel):
     password: SecretStr
     password_bytes: SecretBytes
+    password_idempotent: SecretStr
```
Comment (samuelcolvin) [inline docs/examples/types_secret_types.py:6]: do you need another field to cover the idempotent case?

Surely just a note in the docs should be enough?

File: pydantic/types.py (hunk 1 above)
Comment (samuelcolvin) [reply pydantic/types.py:537]: in case someone uses some subclass in future.

File: tests/test_types.py [hunk 3]
```diff
@@ -1888,6 +1888,14 @@ def test_secretstr_equality():
     assert SecretStr('123') is not SecretStr('123')
 
 
+def test_secretstr_idempotent():
+    class Foobar(BaseModel):
+        password: SecretStr
+
+    # Should not raise an exception
+    _ = Foobar(password=SecretStr('1234'))
```
Comment (samuelcolvin) [inline tests/test_types.py:1896]:
```suggestion
    m = Foobar(password=SecretStr('1234'))
    assert m.password. get_secret_value == '1234'
```

File: docs/examples/types_secret_types.py [hunk 4]
```diff
@@ -3,19 +3,27 @@
...
 
-sm = SimpleModel(password='IAmSensitive', password_bytes=b'IAmSensitiveBytes')
+sm = SimpleModel(password='IAmSensitive', 
+                 password_bytes=b'IAmSensitiveBytes',
```
Comment (samuelcolvin) [inline docs/examples/types_secret_types.py:9]: I think this is not required anymore, but I don't like this format, better to do it the way black suggests if you have to.

File: pydantic/types.py (hunk 1 above)
Comment (Atheuz) [reply pydantic/types.py:537]: I was sure I saw this used somewhere, but I can't find it now, I'll change it to use isinstance instead.

Comment (Atheuz) [reply pydantic/types.py]: This is fixed now.

File: tests/test_types.py (hunk 3 above)
Comment (Atheuz) [reply tests/test_types.py]: This is fixed.

File: docs/examples/types_secret_types.py (hunk 4 above)
Comment (Atheuz) [reply docs/examples/types_secret_types.py]: This is fixed now.

File: docs/examples/types_secret_types.py (hunk 2 above)
Comment (Atheuz) [reply docs/examples/types_secret_types.py]: I added a note in the latest commit and removed this field.

**Pull request merged**